### PR TITLE
[MaintenanceService] Added Support for InGuestPatch maintenance scope.

### PR DESCRIPTION
**Pull request by KalpeshChavan12**
## Description

Added Paramter to cmdlet to create InGuestPatch maintenance config.

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/988 
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by KalpeshChavan12** [comment]
.Net SDK not yet published

https://github.com/Azure/azure-sdk-for-net/pull/23812

Test recorded using local .net sdk build

File: src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs [hunk 1]
```diff
@@ -163,6 +232,55 @@ public override void ExecuteCmdlet()
             HelpMessage = "The Recurrence interval.")]
         public string RecurEvery { get; set; }
 
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "List of packges include during vm patch operation")]
```
Comment (sampreeth1999) [inline src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs:237]: List of **packages to** include during vm patch operation

File: src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs [hunk 2]
```diff
@@ -163,6 +232,55 @@ public override void ExecuteCmdlet()
...
+        public HashSet<string> LinuxParameterPackageNameMaskToInclude { get; set; }
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "List of packages exclude during vm patch operation")]
```
Comment (sampreeth1999) [inline src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs:242]: **packages to exclude**

File: src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs [hunk 3]
```diff
@@ -163,6 +232,55 @@ public override void ExecuteCmdlet()
...
+        public HashSet<string> LinuxParameterPackageNameMaskToExclude { get; set; }
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "List of linux patch classifications")]
+        public HashSet<string> LinuxParameterClassificationToInclude { get; set; }
```
Comment (sampreeth1999) [inline src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs:248]: Should we rename all prefixes to plural?
LinuxParameters
WindowsParameters

File: src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs [hunk 4]
```diff
@@ -163,6 +232,55 @@ public override void ExecuteCmdlet()
...
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "List of KBs include during vm patch operation")]
+        public HashSet<string> WindowParameterKbNumberToInclude { get; set; }
```
Comment (sampreeth1999) [inline src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs:253]: KBs **to** include

File: src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs [hunk 5]
```diff
@@ -163,6 +232,55 @@ public override void ExecuteCmdlet()
...
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "List of KBs excluded during vm patch operation")]
```
Comment (sampreeth1999) [inline src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs:257]:
```suggestion
            HelpMessage = "List of KBs to exclude during vm patch operation")]
```

File: src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs [hunk 6]
```diff
@@ -163,6 +232,55 @@ public override void ExecuteCmdlet()
...
+        public string PreTask { get; set; }
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "List of tasks executed after schedule. [{'source' :'runbook', 'taskScope': 'Resource', 'parameters': { 'arg1': 'value1'}}]")]
+        public string PostTask { get; set; }
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "Install Patch Reboot Option. Allowed values NeverReboot, RebootIfRequired, AlwaysReboot")]
```
Comment (sampreeth1999) [inline src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs:282]:
```suggestion
            HelpMessage = "Install Patch Reboot Option. Allowed values Never, IfRequired, Always")]
```

File: src/Maintenance/Maintenance/help/New-AzMaintenanceConfiguration.md [hunk 7]
```diff
@@ -13,11 +13,14 @@ Create or Update configuration record
...
- [-Tag <Hashtable>] [-ExtensionProperty <Hashtable>] [-MaintenanceScope <String>] [-StartDateTime <String>]
- [-ExpirationDateTime <String>] [-Timezone <String>] [-Duration <TimeSpan>] [-Visibility <String>]
- [-RecurEvery <String>] [-AsJob] [-DefaultProfile <IAzureContextContainer>] [-WhatIf] [-Confirm]
- [<CommonParameters>]
+New-AzMaintenanceConfiguration [-ResourceGroupName] <string> [-Name] <string> [-Location] <string> [-Tag <hashtable>]
+    [-ExtensionProperty <hashtable>] [-MaintenanceScope <string>] [-StartDateTime <string>] [-ExpirationDateTime <string>]
+    [-Timezone <string>] [-Duration <timespan>] [-Visibility <string>] [-RecurEvery <string>]
+    [-LinuxParameterPackageNameMaskToInclude <HashSet[string]>] [-LinuxParameterPackageNameMaskToExclude <HashSet[string]>]
+    [-LinuxParameterClassificationToInclude <HashSet[string]>] [-WindowParameterKbNumberToInclude <HashSet[string]>]
+    [-WindowParameterKbNumberToExclude <HashSet[string]>] [-WindowParameterClassificationToInclude <HashSet[string]>]
```
Comment (sampreeth1999) [inline src/Maintenance/Maintenance/help/New-AzMaintenanceConfiguration.md:21]: Should we create a new command like 
New-InputPatchConfiguration ....

and pass that object to New-AzMaintenanceConfiguration

File: src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs [hunk 8]
```diff
@@ -163,6 +232,55 @@ public override void ExecuteCmdlet()
...
+        public HashSet<string> LinuxParameterPackageNameMaskToExclude { get; set; }
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "List of linux patch classifications")]
+        public HashSet<string> LinuxParameterClassificationToInclude { get; set; }
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "List of KBs include during vm patch operation")]
```
Comment (sampreeth1999) [inline src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs:252]:
```suggestion
            HelpMessage = "List of KBs to include during vm patch operation")]
```

File: src/Maintenance/Maintenance/help/New-AzMaintenanceConfiguration.md [hunk 9]
```diff
@@ -261,6 +354,66 @@ Accept pipeline input: False
 Accept wildcard characters: False
 ```
 
+### -WindowParameterClassificationToInclude
+List of windows patch classification
```
Comment (sampreeth1999) [inline src/Maintenance/Maintenance/help/New-AzMaintenanceConfiguration.md:358]:
```suggestion
List of windows patch classifications
```

File: src/Maintenance/Maintenance/help/New-AzMaintenanceConfiguration.md (hunk 7 above)
Comment (KalpeshChavan12) [reply src/Maintenance/Maintenance/help/New-AzMaintenanceConfiguration.md:21]: That will be two step process for user, here we flatten all parameters,

File: src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs (hunk 3 above)
Comment (KalpeshChavan12) [reply src/Maintenance/Maintenance/MaintenanceConfiguration/MaintenanceConfigurationCreateOrUpdateMethod.cs:248]: As per powershell design its not allowed , please refer design review issue , we should not use plural

File: src/Maintenance/Maintenance/help/New-AzMaintenanceConfiguration.md (hunk 9 above)
Comment (KalpeshChavan12) [reply src/Maintenance/Maintenance/help/New-AzMaintenanceConfiguration.md:358]: As per design we prefer singular

**Comment by dingmeng-xue** [comment]
According to the syntax of `Update-AzMaintenanceConfiguration`, you need to add new parameters to `PSMaintenanceConfiguration`. Otherwise, user cannot update this resource.

**Comment by KalpeshChavan12** [comment]
> According to the syntax of `Update-AzMaintenanceConfiguration`, you need to add new parameters to `PSMaintenanceConfiguration`. Otherwise, user cannot update this resource.

Update accept configuration object as parameter, Use can get object  using Get-AzMaintenanceConfiguration or New-AzMaintennaceConfiguration and mofify object and pass to Update-AzMaintenanceConfiguration.

We need to update object return to user as Get and New cmdlets

Thanks for review.

**Comment by KalpeshChavan12** [comment]
> According to the syntax of `Update-AzMaintenanceConfiguration`, you need to add new parameters to `PSMaintenanceConfiguration`. Otherwise, user cannot update this resource.

Fixed in latest revision.

Please review. PR

**Pull request merged**